OpenRA.Game/Graphics/TerrainSpriteLayer.cs

```diff
@@ -0,0 +1,135 @@
+using OpenRA.Traits;
+
+namespace OpenRA.Graphics
+{
+	public sealed class TerrainSpriteLayer : IDisposable
+	{
+		static readonly Sprite EmptySprite = new Sprite(new Sheet(new Size(1, 1)), Rectangle.Empty, TextureChannel.Alpha);
+
+		readonly IVertexBuffer<Vertex> vertexBuffer;
+		readonly Vertex[] vertices;
+		readonly HashSet<int> dirtyRows = new HashSet<int>();
+		readonly int rowStride;
+
+		readonly WorldRenderer worldRenderer;
+		readonly Map map;
+
+		readonly Sheet sheet;
+		readonly BlendMode blendMode;
+
+		float paletteIndex;
+
+		public TerrainSpriteLayer(World world, WorldRenderer wr, Sheet sheet, BlendMode blendMode, PaletteReference palette)
+		{
+			worldRenderer = wr;
+			this.sheet = sheet;
+			this.blendMode = blendMode;
+			paletteIndex = palette.TextureIndex;
+
+			map = world.Map;
+			rowStride = 4 * map.MapSize.X;
+
+			vertices = new Vertex[rowStride * map.MapSize.Y];
+			vertexBuffer = Game.Renderer.Device.CreateVertexBuffer(vertices.Length);
+
+			wr.PaletteInvalidated += () =>
+			{
+				paletteIndex = palette.TextureIndex;
+
+				// Everything in the layer uses the same palette,
+				// so we can fix the indices in one pass
+				for (var i = 0; i < vertices.Length; i++)
+				{
+					var v = vertices[i];
+					vertices[i] = new Vertex(v.X, v.Y, v.Z, v.U, v.V, paletteIndex, v.C);
+				}
+
+				for (var v = 0; v < map.MapSize.Y; v++)
+					dirtyRows.Add(v);
+			};
+		}
+
+		public void Update(CPos cell, Sprite sprite)
+		{
+			var pos = worldRenderer.ScreenPosition(map.CenterOfCell(cell)) + sprite.Offset - 0.5f * sprite.Size;
+			Update(cell.ToMPos(map.TileShape), sprite, pos);
+		}
+
+		public void Update(MPos uv, Sprite sprite, float2 pos)
+		{
+			if (sprite != null)
+			{
+				if (sprite.Sheet != sheet)
+					throw new InvalidDataException("Attempted to add sprite from a different sheet");
+
+				if (sprite.BlendMode != blendMode)
+					throw new InvalidDataException("Attempted to add sprite with a different blend mode");
+			}
+			else
+				sprite = EmptySprite;
+
+			var offset = rowStride * uv.V + 4 * uv.U;
+			Util.FastCreateQuad(vertices, pos, sprite, paletteIndex, offset, sprite.Size);
+
+			dirtyRows.Add(uv.V);
+		}
+
+		public void Draw(Viewport viewport)
+		{
+			var cells = viewport.VisibleCells;
+
+			// Only draw the rows that are visible.
+			// VisibleCells is clamped to the map, so additional checks are unnecessary
+			var firstRow = cells.TopLeft.ToMPos(map).V;
+			var lastRow = cells.BottomRight.ToMPos(map).V + 1;
+
+			Game.Renderer.Flush();
+
+			// Flush any visible changes to the GPU
+			for (var v = firstRow; v <= lastRow; v++)
+			{
+				if (!dirtyRows.Remove(v))
+					continue;
+
+				var rowOffset = rowStride * v;
+
+				unsafe
+				{
+					// The compiler / language spec won't let us calculate a pointer to
+					// an offset inside a generic array T[], and so we are forced to
+					// calculate the start-of-row pointer here to pass in to SetData.
+					fixed (Vertex* vPtr = &vertices[0])
```

Comment:
Indeed.  It looks like you originally set it up that way to avoid a cycling dependency between `Vertex` (in Game.exe) and `IVertexBuffer` implementations in the renderer dlls.